Dont heal event references

mplorentz opened the pull request:
## Issues covered
This is a solution to problem 1 in https://github.com/planetary-social/nos/issues/703#issuecomment-2147829931
> The database cleanup script is deleting EventReferences and replacing them with new ones, breaking SwiftUI observation. Solution: refactor EventReferences out of model layer.

I got this wrong in my initial assessment. It wasn't `EventReference`s that were being deleted and recreated, it was the `Event`s themselves. But the fix was similar, see below.

## Description
The database cleanup routine is coded to clear out old `Events`. This created a problem where an `Event` that wasn't cleared out could reference an older `Event` that _was_ cleared out. The old solution to this problem was to "heal" these broken references by recreating the referenced events using `Event.findOrCreateStub(by:...)`. This wasn't good enough. 

Because we were deleting and recreating events with the same `identifier` in the cleanup routine they don't get created with the same Core Data `objectID` and Core data treats them as new object. This confuses the SwiftUI observation code and if it was observing the original event it won't detect updates to the new one until the view is recreated.

The new code is smarter and starts by deleting old `EventReference`s first. Then it only deletes `Event`s that aren't referenced by any other event.

~~I made another change which isn't really related to the problem but I wanted to give it a try anyway. We had some complex logic to try to keep 1000 recent events in the db and just delete older ones. But why? They are old and you have probably already read them. We are already fetching a bunch of new data when the app opens. Given that we have a database size problem lurking I decided to just delete all old events rather than trying to keep 1000. It simplified the script quite a bit. We actually do keep all old events of the logged in user, and events for stories they have recently read, and under some other conditions, but we are no longer arbitrarily keeping 1000 others around.~~

## How to test
This should practically eliminate infinite spinners on reposts. Please scroll around the feeds and look for some. The only time a spinner should show infinitely is if 1) the source event does **not** contain the reposted event in its `content` field and 2) the reposted event is not present on any of the user's relays.

mplorentz commented on NosTests/Model/AuthorTests.swift:
```diff
@@ -226,26 +226,27 @@ final class AuthorTests: CoreDataTestCase {
-        event.allTags = tags as NSObject
-        return event
+    
+    func test_outOfNetwork_givenCircleOfFollows() throws {
```
This is a little bonus test I threw in because I had a hunch that the cleanup function was deleting too many Follows or Authors, so I wrote a quick test for that. It looks like it is functioning as intended though.

mplorentz commented on NosTests/Model/EventTests.swift:
```diff
@@ -39,7 +43,21 @@ final class EventTests: CoreDataTestCase {
+        )
+    }
+
+    func testIdentifierCalculationWithEmptyTags() throws {
```
I found a bug when creating the `EventFixture` where an `Event` with no tags and one with an empty array of tags would compute different identifiers. So I fixed that and added a test for it.

martindsq replied:
I think it would be more communicative (and self documented) if the test includes XCTAssertNotEqual to an identifier generated by a event without tags.

mplorentz replied:
Ok, I started to do that but it made more sense to me to just have two XCTAssertEquals with different values next to them. Let me know what you think.

mplorentz commented on NosTests/Service/FileStorage/NostrBuildAPIClientTests.swift:
```diff
@@ -107,7 +107,7 @@ class NostrBuildAPIClientTests: XCTestCase {
 
         // Act
-        let (uploadRequest, data) = try subject.uploadRequest(fileAt: fileURL)
+        let (uploadRequest, _) = try subject.uploadRequest(fileAt: fileURL)
```
This fixes a build warning that was introduced recently.

martindsq commented on Nos/Controller/PersistenceController.swift:
```diff
@@ -198,7 +198,9 @@ class PersistenceController {
         do {
-            try await DatabaseCleaner.cleanupEntities(before: Date.now, for: authorKey, in: context)
+            // We don't want to delete any events downloaded after app boot, so we subtract 60 seconds 
+            let cleanupBeforeDate = Date.now.addingTimeInterval(-60) // Subtract 60 seconds
```
You can just do `let cleanupBeforeDate = Date.now - 60 // Subtract 60 seconds`

mplorentz replied:
🤯

joshuatbrown replied:
🤯

martindsq commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -342,6 +342,37 @@ public class Event: NosManagedObject, VerifiableEvent {
         return fetchRequest
     }
     
+    /// A fetch requests for all the events that should be cleared out of the database by 
```
```suggestion
    /// A fetch request for all the events that should be cleared out of the database by 
```

martindsq commented:
> I made another change which isn't really related to the problem but I wanted to give it a try anyway. We had some complex logic to try to keep 1000 recent events in the db and just delete older ones. But why? They are old and you have probably already read them. We are already fetching a bunch of new data when the app opens. Given that we have a database size problem lurking I decided to just delete all old events rather than trying to keep 1000.

This has a disadvantage I'd like to consider before continuing. If you close the app and try to open the app without Internet, you just lose the posts you had (except the ones you posted) because instead of keeping the recent 1000, we just delete all of them in hope we can just fetch them from a relay.

I can think a few cases where this could be disruptive, imagine you want to open the app to re-read a post you kept to be read later. I guess we can add an option to "save" posts (which could be useful by itself alone). But I wanted to bring this issue anyway so we are conscious about all of the consequences.

martindsq commented on Nos/Service/DatabaseCleaner.swift:
```diff
@@ -5,8 +5,6 @@ import Dependencies
 
 enum DatabaseCleaner {
     
-    // swiftlint:disable function_body_length 
```
👍

martindsq reviewed (commented):
The code is looking good. Thanks for the unit tests you added, I know you wanted to write even more, but I already appreciate the ones you wrote. I left some minor comments.

rabble commented:
Is this deployed in staging build 320?

mplorentz commented:
> Is this deployed in staging build 320?

No, PRs only go to staging after they are merged.

mplorentz commented:
> If you close the app and try to open the app without Internet, you just lose the posts you had (except the ones you posted) because instead of keeping the recent 1000, we just delete all of them in hope we can just fetch them from a relay.

@martindsq It's a good point. I'm don't think many people use app offline and it's never a feature we have committed to. It's just a side-effect of the way we designed things. But I think the main reason I wanted to nuke this logic when I started working on this PR is because I wasn't trusting that it worked correctly. Now that the PR is done and we have some tests I feel like I could trust it. I'll try adding it back with some tests.

martindsq commented:
> > If you close the app and try to open the app without Internet, you just lose the posts you had (except the ones you posted) because instead of keeping the recent 1000, we just delete all of them in hope we can just fetch them from a relay.
> 
> @martindsq It's a good point. I'm don't think many people use app offline and it's never a feature we have committed to. It's just a side-effect of the way we designed things. But I think the main reason I wanted to nuke this logic when I started working on this PR is because I wasn't trusting that it worked correctly. Now that the PR is done and we have some tests I feel like I could trust it. I'll try adding it back with some tests.

It's not the offline use they way some people used scuttlebutt that I'm worried about. But people with intermittent connection, like when in a subway, in a remote place like countryside, or in an event, like dweb camp. I have fixed some bugs that we didn't spot before because they are mostly seen when having a bad network, and that can impact too if the relays are slow to respond. Can we put Linda aware of this? I know she has bad network sometimes and also she is aware on the network conditions some of the content creators have. If she is ok with this, then we can merge.

mplorentz commented:
@martindsq I had already added back our "keep 1000 events" logic before I saw your last comment about consulting Linda. It didn't take very long to add it back, so now we are back to no behavior changes there. The user should still be able to fire up the app in no or low-connectivity and see the stuff they were looking at when the app was last open.

However I found another problem with my previous logic that was very tricky to solve. My code to just delete EventReferences where both events were old was not enough because we save some old events (mainly the ones published by the current user). So deleting event references used by those old events would break the links on them too. 

The solution I came up with was to add an exception to only delete EventReferences where both events were old **and** not "protected" events that we are going to save forever. This presented an additional problem where chains of old events referencing other old events could keep way more than 1000 events around. So I added a routine that converts some old events back into stubs - in particular the old events that are referenced by newer or "protected" events but otherwise we don't need to keep around.

So that's why it took me so long to get this ready for review again, but here it is now.

martindsq commented on Nos/Service/DatabaseCleaner.swift:
```diff
@@ -122,5 +78,79 @@ enum DatabaseCleaner {
+    ) throws {
+        let request = NSFetchRequest<Event>(entityName: "Event")
+        request.sortDescriptors = [NSSortDescriptor(keyPath: \Event.receivedAt, ascending: true)]
+        let oldEventPredicate = NSPredicate(format: "(receivedAt < %@)", deleteBefore as CVarArg)
```
Do you think this predicate and the one below could be shared with the Event class (oldEventClause) so that we have just one NSCompoundPredicate(andPredicateWithSubpredicates) that uses oldEvent and referencedEvents.count > 0 as subpredicates and can be shared in both places? (I'm thinking about a public function that builds this predicate) While I think it doesn't do any magic, the code would be just a bit more organized, but if you think this would be over-engineering this, let me know, the code looks good enough to me.

martindsq replied:
Very clear. Thanks for giving it a shot.

martindsq commented on Nos/Models/CoreData/Event+CoreDataClass.swift:
```diff
@@ -342,6 +345,112 @@ public class Event: NosManagedObject, VerifiableEvent {
+    /// - Parameter before: The date before which events will be considered for cleanup. 
+    /// - Parameter user: The Author record for the currently logged in user. Special treatment is given to their data.
+    @nonobjc public class func protectedFromCleanupPredicate(for user: Author) -> NSPredicate {
+        // NOTE: This code is pretty much the same as that of `protectedFromCleanupSubqueryPredicate` but I can't 
```
One way would be to have one function, but prepend `$event.` in one case but not in the other, for example, with a `isSubquery` argument. I think it doesn't look good nor polished, but at least it would be just one function. Maybe @joshuatbrown can provide another idea here.

joshuatbrown replied:
Yeah, I think your suggestion is good. I don't have a better idea at this point, but that's a straightforward, simple way to avoid needing to update both. Seems better so these don't get out of sync. Something like:

```
func protectedFromCleanupPredicate(for user: Author, isSubquery: Bool) {
    guard...
    let eventOrEmpty = isSubquery ? "$event." : "" // or let prefix, maybe?
    let userEventsPredicate = NSPredicate(format: "\(eventOrEmpty)author.hexadecimalPublicKey = '\(userKey)'")
    ...
}
```

I needed a visual to see how it would -- that seems fine to me, I think, and probably better than two methods.

martindsq replied:
Awesome. Great suggestions as usual, Josh!

joshuatbrown replied:
@martindsq this was your idea! You get the credit! 😀

martindsq reviewed (changes requested):
Thanks for the in-depth comments you added. This is looking very good. I just added some small change requests, but we are nearly there with this PR.

mplorentz commented on NosTests/Service/DatabaseCleanerTests.swift:
```diff
@@ -20,10 +20,10 @@ final class DatabaseCleanerTests: CoreDataTestCase {
     // MARK: - EventReferences
     
-    func test_cleanup_deletesCorrectEventReferences() async throws {
+    @MainActor func test_cleanup_deletesCorrectEventReferences() async throws {
```
Because `testContext` from the superclass is actually the `viewContext` which is bound to the main thread. When you mark a test function `async` then it sometimes gets executed on a background thread, which sometimes Core Data doesn't like and it will cause a test failure.

I tried adding @MainActor to the whole class but got some warning about actor conformance being different from the superclass or something, so I just added the attribute to each function instead.

martindsq replied:
Good to know. I assumed the unit tests always ran in the "main" thread given that there is no limits when testing. But it makes sense that they run in the thread they want.

martindsq reviewed (approved):
I tested it again and didn't see any issue. We can merge this already.

Good work!

mplorentz commented:
Hm, I saw a DatabaseCleanerTest fail in CI but it passed 100 times locally for me 🤔 I wonder if #1253 is to blame. I'm going to rerun the tests and this should merge if they pass, but I'll keep a close eye out for further issues with them.

mplorentz commented:
Hm, looks like a different test is failing every time.

mplorentz commented:
Ok a different test is failing every time, so I'm going to force merge and we'll fix the test issues in #1255.
